File: Avg.py
from Tuple import Tuple
from Instrumentation import Instrumentation
from Operateur import Operateur

class Avg(Instrumentation, Operateur):

    def __init__(self, _in, _col):
        super().__init__("Avg" + str(Instrumentation.number))
        Instrumentation.number += 1
        self.start()
        self.col = _col
        self.child = _in
        self.stop()

    def open(self):
        self.start()
        self.child.open()
        self.tuplesProduits = 0
        self.memoire = 0
        
        first = self.child.next()
        if first is not None:
            count = 1
            c = self.col
            sum = first.val[c]
            while True:
                temp = self.child.next()
                if temp is None:
                    break
                sum += temp.val[c]
                count += 1
            self.tempValAvg.val[c] = sum / count
        self.stop()

    def next(self):
        self.start()
        if self.tempValAvg is None:
            self.stop()
            return None
        else:
            ret = Tuple(1)
            ret.val[0] = self.tempValAvg.val[self.col]
            self.tempValAvg = None
            self.produit(ret)
            self.stop()
            return ret
    
    def close(self):
        self.child.close()

File: Avg.java

public class Avg extends Instrumentation implements Operateur {

    private int col;
    private Operateur in; 
    private Tuple tempValAvg;
    private int count;

    public Avg(Operateur _in, int _col) {
        super("AVG" + Instrumentation.number++);
        this.start();
        this.col = _col;
        this.in = _in;
        this.tempValAvg = null;
        this.stop();
    }

    @Override
    public void open() {
        this.start();
        this.in.open();
        this.tuplesProduits = 0;
        this.memoire = 0;
        Tuple temp = null;
        this.tempValAvg = this.in.next();
        if (this.tempValAvg != null) {
            this.count = 1;
            while ((temp = this.in.next()) != null) {
                count++;
                this.tempValAvg.val[this.col] += temp.val[this.col];
            }
            this.tempValAvg.val[this.col] /= count;
        }
        this.stop();
    }

    @Override
    public Tuple next() {
        this.start();
        if (this.tempValAvg == null) {
            this.stop();
            return null;
        } else {
            Tuple ret = new Tuple(1);
            ret.val[0] = this.tempValAvg.val[this.col];
            this.tempValAvg = null;
            this.produit(ret);
            this.stop();
            return ret;
        }
    }

    @Override
    public void close() {
        this.in.close();
    }

}
